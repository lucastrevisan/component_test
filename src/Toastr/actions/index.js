export default () => ({
  openToastr: (_, config) => {
    return { toastrSettings: { isOpen: true, ...config } };
  },
  closeToastr: () => ({ toastrSettings: false })
});
