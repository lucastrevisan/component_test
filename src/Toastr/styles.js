import styled, { css } from "styled-components";
import { lightGreen, white } from "../Common/styles/settings";
import { rem, media } from "../Common/styles/tools";

export const Wrapper = styled.div`
  cursor: pointer;
  position: fixed;
  opacity: 0;
  top: -${rem(50)};
  z-index: 1;
  width: 100%;
  min-height: ${rem(34)};
  background-color: ${lightGreen};
  box-shadow: 0 ${rem(4)} ${rem(22)} 0 rgba(28, 205, 163, 0.3);
  transition: 0.4s ease-in-out;

  ${media.md(css`
    min-height: ${rem(40)};
  `)};

  :hover {
    opacity: 0.8;
  }

  ${({ opened }) =>
    opened &&
    css`
      opacity: 1;
      top: 0;
    `};

  ${({ opened }) =>
    !opened &&
    css`
      opacity: 0;
      top: -${rem(50)};
    `};
`;

export const Text = styled.p`
  font-size: ${rem(14)};
  font-family: "Lato";
  letter-spacing: ${rem(0.1)};
  text-align: center;
  padding: ${rem(10)} 0 ${rem(5)};
  color: ${white};
  width: 90%;
  margin: 0 auto;

  ${media.md(css`
    font-size: ${rem(15)};
    padding-top: ${rem(12)};
  `)};
`;

export const IconWrapper = styled.div`
  position: absolute;
  top: ${rem(10)};
  right: ${rem(10)};

  ${media.md(css`
    top: ${rem(12)};
  `)};
`;
