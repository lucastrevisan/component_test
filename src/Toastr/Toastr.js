import React, { Component } from "react";
import { string, shape, func, bool, oneOfType, object } from "prop-types";

import Icon from "../Icon";
import { Wrapper, Text, IconWrapper } from "./styles.js";
import { white } from "../Common/styles/settings";

class Toastr extends Component {
  constructor(props) {
    super(props);

    this.state = {
      mounted: false
    };

    this.toggle = this.toggle.bind(this);
  }

  toggle() {
    setTimeout(() => {
      this.setState({ mounted: true });
    }, 500);
  }

  componentWillMount() {
    this.historyListenner = this.props.history.listen(this.props.closeToastr);
  }

  componentDidMount() {
    this.toggle();
  }

  componentWillUnmount() {
    this.historyListenner();
  }

  render() {
    const { toastrSettings, closeToastr } = this.props;
    const { mounted } = this.state;

    return (
      <Wrapper
        data-test="Toastr"
        opened={mounted && toastrSettings}
        onClick={closeToastr}
      >
        <Text>{toastrSettings && toastrSettings.text}</Text>
        <IconWrapper data-test="ToastrClose">
          <Icon type="ToastrClose" color={white} />
        </IconWrapper>
      </Wrapper>
    );
  }
}

Toastr.defaultProps = {
  toastrSettings: false
};

Toastr.propTypes = {
  toastrSettings: oneOfType([
    bool,
    shape({
      text: string
    }).isRequired
  ]),
  history: shape({
    location: object,
    listen: func
  }).isRequired,
  closeToastr: func.isRequired
};

export default Toastr;
