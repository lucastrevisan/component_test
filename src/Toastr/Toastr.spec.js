import React from "react";
import { shallow } from "enzyme";

import Toastr from "./Toastr";

const listenMock = jest.fn();

const history = {
  listen: listenMock
};

const closeToastr = listenMock;

describe("Toastr component", () => {
  it("should match snapshot with toastrSettings as false", () => {
    expect(
      shallow(<Toastr toastrSettings={false} history={history} closeToastr={closeToastr} />)
    ).toMatchSnapshot();
  });

  it("should match snapshot with toastrSettings as an object", () => {
    expect(
      shallow(<Toastr toastrSettings={{ text: "text" }} history={history} closeToastr={closeToastr} />)
    ).toMatchSnapshot();
  });

  it("should invoke listen function", () => {
    shallow(<Toastr toastrSettings={{ text: "text" }} history={history} closeToastr={closeToastr} />);
    expect(listenMock).toHaveBeenCalled();
  });

  describe("toggle", () => {
    it("shouldn't update the state before the timeout", () => {
      const instance = shallow(
        <Toastr toastrSettings={{ text: "text" }} history={history} closeToastr={closeToastr} />
      ).instance();

      instance.toggle();

      expect(instance.state.mounted).toBe(false);
    });

    it("should update the state after the timeout", done => {
      const instance = shallow(
        <Toastr toastrSettings={{ text: "text" }} history={history} closeToastr={closeToastr} />
      ).instance();

      instance.toggle();

      setTimeout(() => {
        expect(instance.state.mounted).toBe(true);
        done();
      }, 500);
    });
  });
});
