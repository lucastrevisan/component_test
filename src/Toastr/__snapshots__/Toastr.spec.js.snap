// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toastr component should match snapshot with toastrSettings as an object 1`] = `
<styled.div
  data-test="Toastr"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [
          [MockFunction] {
            "calls": [Circular],
          },
        ],
        Array [
          [MockFunction] {
            "calls": [Circular],
          },
        ],
      ],
    }
  }
  opened={false}
>
  <styled.p>
    text
  </styled.p>
  <styled.div
    data-test="ToastrClose"
  >
    <Icon
      color="#ffffff"
      height={null}
      type="ToastrClose"
      width={null}
    />
  </styled.div>
</styled.div>
`;

exports[`Toastr component should match snapshot with toastrSettings as false 1`] = `
<styled.div
  data-test="Toastr"
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [
          [MockFunction] {
            "calls": [Circular],
          },
        ],
      ],
    }
  }
  opened={false}
>
  <styled.p />
  <styled.div
    data-test="ToastrClose"
  >
    <Icon
      color="#ffffff"
      height={null}
      type="ToastrClose"
      width={null}
    />
  </styled.div>
</styled.div>
`;
